### PR TITLE
Sample weights feature selection 

### PR DESCRIPTION
**Pull request by harjeetkalsi**

### CODE REVIEW

File: mlxtend/feature_selection/sequential_feature_selector.py [hunk 1]
```diff
@@ -14,25 +14,73 @@
...
 from sklearn.base import MetaEstimatorMixin
 from ..externals.name_estimators import _name_estimators
 from ..utils.base_compostion import _BaseXComposition
-from sklearn.model_selection import cross_val_score
+from sklearn.model_selection import StratifiedKFold, cross_val_score
 from joblib import Parallel, delayed
 
 
-def _calc_score(selector, X, y, indices, groups=None, **fit_params):
+def fit_and_score(model, X, y, weights,
```
Comment (philipp-ludersdorfer) [inline mlxtend/feature_selection/sequential_feature_selector.py]: It feels like this and `parallel_cross_val_scores_weighted()` should be hidden functions (i.e. `_fit_and_score`). What do you think?

File: mlxtend/feature_selection/sequential_feature_selector.py [hunk 2]
```diff
@@ -14,25 +14,73 @@
 import sys
 from copy import deepcopy
 from itertools import combinations
-from sklearn.metrics import get_scorer
+from sklearn.metrics import get_scorer, roc_auc_score, accuracy_score
```
Comment (philipp-ludersdorfer) [inline mlxtend/feature_selection/sequential_feature_selector.py]: I don't think we need to import roc_auc_score and accuracy_score. 

Sorry, I think this is completely my fault. I didn't make the arguments in the functions I used in my example notebook generic enough.

I have tried clarified my thoughts regarding the scoring metric below.

File: mlxtend/feature_selection/sequential_feature_selector.py [hunk 3]
```diff
@@ -14,25 +14,73 @@
...
 from ..utils.base_compostion import _BaseXComposition
-from sklearn.model_selection import cross_val_score
+from sklearn.model_selection import StratifiedKFold, cross_val_score
 from joblib import Parallel, delayed
 
 
-def _calc_score(selector, X, y, indices, groups=None, **fit_params):
+def fit_and_score(model, X, y, weights,
+                  train, test,
+                  metric=roc_auc_score):
```
Comment (philipp-ludersdorfer) [inline mlxtend/feature_selection/sequential_feature_selector.py]: I think we should rename `metric` to `scoring` and don't assign a default here. 
It will only be used by the SequentialFeatureSelector class which passes its .scorer attribute to the function anyway.
Then we also don't need to import roc_auc_score (and accuracy_score) at the top.

File: mlxtend/feature_selection/sequential_feature_selector.py [hunk 4]
```diff
@@ -14,25 +14,73 @@
...
+    model_clone = clone(model)
+    model_clone.fit(X[train], y[train], sample_weight=weights[train])
+    y_pred = model_clone.predict(X[test])
+
+    # Score
+    score = metric(y[test], y_pred, sample_weight=weights[test])
+    return score
+
+
+def parallel_cross_val_scores_weighted(model, X, y, weights,
```
Comment (philipp-ludersdorfer) [inline mlxtend/feature_selection/sequential_feature_selector.py]: I don't think the 'parallel' in the function name is necessary. Everything is parallel these days 😬

File: mlxtend/feature_selection/sequential_feature_selector.py [hunk 5]
```diff
@@ -14,25 +14,73 @@
...
+                                       cv=5,
+                                       metric=accuracy_score,
```
Comment (philipp-ludersdorfer) [inline mlxtend/feature_selection/sequential_feature_selector.py]: As above: I think we should rename it to scoring and don't have a default.

File: mlxtend/feature_selection/sequential_feature_selector.py [hunk 6]
```diff
@@ -14,25 +14,73 @@
...
+                                       n_jobs=None,
+                                       verbose=0,
+                                       pre_dispatch='2*n_jobs'):
+
+    print("here")
+    print(weights)
```
Comment (philipp-ludersdorfer) [inline mlxtend/feature_selection/sequential_feature_selector.py]: Do we nee these two print statements?

File: mlxtend/feature_selection/sequential_feature_selector.py (hunk 1 above)
Comment (philipp-ludersdorfer) [inline mlxtend/feature_selection/sequential_feature_selector.py]: Also, just to be super clear what we mean - can we change the `weights` argument to `sample_weights`?

File: mlxtend/feature_selection/sequential_feature_selector.py (hunk 4 above)
Comment (philipp-ludersdorfer) [inline mlxtend/feature_selection/sequential_feature_selector.py]: Same as above: `sample_weights` instead of `weights`

File: mlxtend/feature_selection/sequential_feature_selector.py [hunk 7]
```diff
@@ -14,25 +14,73 @@
...
+
+    # Call fit_and_score
+    scores = parallel(
+        delayed(fit_and_score)(model, X, y, weights, train, test, metric)
+        for train, test in cv.split(X, y))
+
+    return scores
+
+
+def _calc_score(selector, X, y, indices, groups=None, weights=None, **fit_params):
```
Comment (philipp-ludersdorfer) [inline mlxtend/feature_selection/sequential_feature_selector.py]: I don't think you need `weights` here as you also defined the sample weights as an attribute of `selector`.

File: mlxtend/feature_selection/sequential_feature_selector.py [hunk 8]
```diff
@@ -14,25 +14,73 @@
...
-                                 n_jobs=1,
-                                 pre_dispatch=selector.pre_dispatch,
-                                 fit_params=fit_params)
+        if weights is not None:
+            scores = parallel_cross_val_scores_weighted(selector.est_,
+                                     X[:, indices], y,
+                                     weights,
+                                     cv=selector.cv,
+                                     n_jobs=1,
+                                     pre_dispatch=selector.pre_dispatch,
```
Comment (philipp-ludersdorfer) [inline mlxtend/feature_selection/sequential_feature_selector.py]: We need to add `scoring=selector.scorer`* here otherwise it won't pass the correct scoring method .
(*this is assuming that you have changed `metric` to `scoring` in the functions above)

File: mlxtend/feature_selection/sequential_feature_selector.py [hunk 9]
```diff
@@ -14,25 +14,73 @@
...
-                                 groups=groups,
-                                 cv=selector.cv,
-                                 scoring=selector.scorer,
-                                 n_jobs=1,
-                                 pre_dispatch=selector.pre_dispatch,
-                                 fit_params=fit_params)
+        if weights is not None:
+            scores = parallel_cross_val_scores_weighted(selector.est_,
+                                     X[:, indices], y,
+                                     weights,
```
Comment (philipp-ludersdorfer) [inline mlxtend/feature_selection/sequential_feature_selector.py]: If you agree with my comment above, this can be changed to`sample_weights=selector.sample_weights`

File: mlxtend/feature_selection/sequential_feature_selector.py [hunk 10]
```diff
@@ -399,7 +449,7 @@ def fit(self, X, y, custom_feature_names=None, groups=None, **fit_params):
                 k_idx = self.fixed_features_
                 k = len(k_idx)
                 k_idx, k_score = _calc_score(self, X_, y, k_idx,
-                                             groups=groups, **fit_params)
+                                             groups=groups, weights=self.sample_weights, **fit_params)
```
Comment (philipp-ludersdorfer) [inline mlxtend/feature_selection/sequential_feature_selector.py]: See my comment above. I don't think you need `weights` here since you have `self`.

**Review by philipp-ludersdorfer — CHANGES_REQUESTED** [review]
I have a few small comments (most of them are nitpicks 😬 ). 
See what you think.

**Comment by MaxAwan** [comment]
Had a look through, and the only things I could notice were those which @philipp-ludersdorfer has already commented on. Looks good to me after his requested changes are addressed :)

File: mlxtend/feature_selection/sequential_feature_selector.py (hunk 6 above)
Comment (harjeetkalsi) [reply mlxtend/feature_selection/sequential_feature_selector.py]: ah good spot i left them in there!
